languages/swift/exercises/concept/loops/.docs/instructions.md

```diff
@@ -0,0 +1,85 @@
+
+Your first task is to implement the function `hashIDs(\_:) which, given an array of ID numbers, takes each ID, transforms it with Santa's magic hash, and constructs a tuple of the original ID and the hashed ID, returning an array of these tuples.
+
+Santa's magic hash is implemented by multiplying the ID by the magic constant _p_ and taking the remainder modulo another magic constant, _m_. This year, you are given the following magic constants:
+
+- m = `3_999_999_979`
+- p = `2_305_843_009`
+
+```swift
+let ids = [980035519, 1402277754, 1821223877]
+hashIDs(ids)
+// => [(980035519, 1921828986), (1402277754, 712641739), (1821223877, 2651471804)]
+```
+
+## 2. Compute the digital sum of a number
+
+In order to extract Santa's ranking from a hashed ID, you will need to first write a helper function that computes the digital sum of a number. The digital sum of a number is just what it sounds like, the sum of the digits of a number. E.g. the digital sum of 10437 is 1 + 0 + 4 + 3 + 7 = 15.
+
+For this task you will need to implement the function `digitalSum(_:)` that takes an `Int` as input and returns an `Int`. If the input number is less than 0, the function should return 0, otherwise the digital sum of the input as an `Int`.
+
+```swift
+digitalSum(10437)
+// => 15
+digitalSum(992)
+// => 20
+digitalSum(-111)
+// => 0
+```
+
+## 3. Compute the ranking level for a hashed ID
+
+For this task, you will need to implement the function, `rankingLevel(_:)` that takes a hashed ID as input and computes the ranking level for the given input in terms of what is known as a Harshad number.
+
+A Harshad number is simply a number that is evenly divisible by the sum of its digits. So 1475 is not a Harshad number because 1 + 4 + 7 + 5 = 17, and 1475 is not divisible by 17. However, 8118 is a Harshad number because 8 + 1 + 1 + 8 = 18 and 8118 is divisible by 18.
+
+Given this definition of a Harshad number, the ranking level of a number is the length of the sequence of _distinct_ Harshad numbers where each element of the sequence is equal to the previous element divided by the sum of its digits.
+
+For example, consider the numbers 5217, 8118, and 7560:
+
+5217 has a digital sum of 5 + 2 + 1 + 7 = 15 but 5217 is not a multiple of 15, so the sequence is empty and the ranking level of 5217 is 0.
+
+8118 has a digital sum of 8 + 1 + 1 + 8 = 18, and 8118 / 18 = 451.  
+451 has a digital sum of 4 + 5 + 1 = 10, but 451 is not a multiple of 10, so we stop the process.
+
+Here, our sequence is just [8118]. Its length is 1 and so the ranking level is 1.
+
+7560 has a digital sum = 7 + 5 + 6 + 0 = 18, and 7560 / 18 => 420.  
+420 has a digital sum of 4 + 2 + 0 = 6 and 420 / 6 => 70.  
+70 has a digital sum of 7 + 0 = 7 and 70 / 7 => 10.  
+10 has a digital sum of 1 + 0 = 1 and 10 / 1 => 10, but this is not a distinct number, it is already in our sequence, so we stop the process.
+
+Here our sequence is [7560, 420, 70, 10]. It's length is 4 and thus our ranking level is 4.
```

Comment:
On the one hand I really like this problem, but I feel like this type of problem (an algorithmic one) is perhaps better suitable for a practice exercise. While I feel it is fun to work on, it took be some time to fully understand what was being asked of me. In general, we want the concept exercises to be easy to understand

According to [the specs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md):

> A seasoned developer in Language X should be able to work through all the Concept Exercises on that track spending no more than 5-10 minutes solving each one.

I think this task might be stretching it a bit. Perhaps you could come up with a slightly simpler task?

p.s. we've recently started encouraging people to submit draft PR's with just three or four files:

- A `design.md` document
- An `instructions.md` document
- An example implementation
- An `introduction.md` document (optional)

The idea of these draft PR's is to get quick feedback on the direction an exercise is going in. This includes the complexity of the tasks, the validity of the story, the scope of the exercise, and such. It might be worth considering doing these draft PR's to lessen the chance of having to re-do work :)

Reply:
@wneumann Thoughts on this?

Reply:
Yeah. I really like the problem, but it is harder to explain than it is to actually solve, so it's probably better for a practice exercise. I'll work on a new exercise.

As for the draft PR, the main problem is that the instructions and example implementation are the last things I write.but I can try reversing that for future concept exercises.

Reply:
> Yeah. I really like the problem, but it is harder to explain than it is to actually solve, so it's probably better for a practice exercise. I'll work on a new exercise.

Brilliant. Make sure to keep this version of the exercise somewhere safe to later be able to convert it to a practice exercise.

> As for the draft PR, the main problem is that the instructions and example implementation are the last things I write.but I can try reversing that for future concept exercises.

What is your usual order creating files for a Concept Exercise? The draft suggestion is just that: a suggestion. I mostly suggested this as I can feel it is not ideal when you have to change an exercise you've already invested quite some time in. A draft could possibly allow us to more quickly give feedback.

Reply:
> What is your usual order creating files for a Concept Exercise? 

I usually start with the design to get the points to cover down, then I do the after to write out everything that needs to be covered. Then I edit it down for the introduction to remove a bunch of the details that aren't needed. Form there I do the introduction, the example and tests, and then the hints. But if there is an existing story for me to poach, then I can do the exercise first -- it's just that coming up with a good, non-contrived exercise with a good story that hits all the important points is tough.

Though I'm learning that a part of that is that I'm trying to pack too much into one exercise. I need to break some into smaller chunks and separate them into multiple concept exercises.

Reply:
> it's just that coming up with a good, non-contrived exercise with a good story that hits all the important points is tough.

Yeah, this is definitely hard. Our [list of stories](https://github.com/exercism/v3/blob/master/reference/stories/README.md#stories) can possibly help here.

> Though I'm learning that a part of that is that I'm trying to pack too much into one exercise. I need to break some into smaller chunks and separate them into multiple concept exercises.

👍 This is definitely something that is different in v3, where exercises are quite small and very focused.